source/common/upstream/eds.cc

```diff
@@ -13,6 +13,30 @@
 
 namespace Envoy {
 namespace Upstream {
+namespace {
+
+HostMetadata parseEndpointMetadata(const envoy::api::v2::LbEndpoint& endpoint) {
+  HostMetadata ret;
+  if (endpoint.has_metadata()) {
+    const auto& filter_metadata =
+        endpoint.metadata().filter_metadata().find(Envoy::Config::MetadataFilters::get().ENVOY_LB);
+    if (filter_metadata == endpoint.metadata().filter_metadata().end()) {
+      return ret;
+    }
+
+    for (auto it : filter_metadata->second.fields()) {
+      if (it.second.kind_case() == ProtobufWkt::Value::kStringValue) {
+        ret[it.first] = it.second.string_value();
+      } else if (it.second.kind_case() == ProtobufWkt::Value::kBoolValue &&
+                 it.second.bool_value()) {
+        ret[it.first] = "true";
+      }
+    }
+  }
+  return ret;
```

Comment:
Not having a complete end-to-end picture of how this will work, I'm not sure if I see the rationale between going from a typed object, which already exists in the proto, to an untyped string map. Why can't HostMetadata just be the https://github.com/lyft/envoy-api/blob/master/api/eds.proto#L74 Metadata message object itself?

Reply:
My thought was to strip out one layer of indirection since the Metadata message is effectively a map of string to map of string to ProtobufWkt::Value. But perhaps I misunderstood its purpose and we actually want to keep the outer map.

Reply:
And the reduction to strings was an attempt to simplify usage, but I don't have strong feelings about that. I expect most of the keys will be strings or booleans (e.g. present or not-present).

Reply:
I'm with @htuch here. IMO we should just expose out the raw metadata object, since it's a map of string -> struct, there is no reason that one of the values might not be fully structured data that could be read into a proto. I'm also not crazy about how we have moved from canary() being a simple var lookup to now requiring a map lookup. Personally I would just leave that is is. (Basically if we want to special case some metadata that's fine).